File: src/core/data_formats/parsing/unitxt.ts
import { prs_decompress } from "../compression/prs/decompress";
import { Cursor } from "../block/cursor/Cursor";
import { Result, success } from "../../Result";

export type Unitxt = string[][];

export function parse_unitxt(buf: Cursor, compressed: boolean = true): Result<Unitxt> {
    if (compressed) {
        const decompression_result = prs_decompress(buf);

        if (!decompression_result.success) {
            return decompression_result;
        }

        buf = decompression_result.value;
    }

    const category_count = buf.u32();
    const entry_counts = buf.u32_array(category_count);
    const category_entry_offsets: number[][] = [];

    for (const entry_count of entry_counts) {
        category_entry_offsets.push(buf.u32_array(entry_count));
    }

    const categories: Unitxt = [];

    for (const category_entry_offset of category_entry_offsets) {
        const entries: string[] = [];
        categories.push(entries);

        for (const entry_offset of category_entry_offset) {
            buf.seek_start(entry_offset);
            const str = buf.string_utf16(Math.min(1024, buf.bytes_left), true, true);
            entries.push(str);
        }
    }

    return success(categories);
}

File: lib/src/commonMain/kotlin/world/phantasmal/lib/fileFormats/Unitxt.kt
package world.phantasmal.lib.fileFormats

import world.phantasmal.lib.cursor.Cursor
import kotlin.math.min

class Unitxt(val categories: List<List<String>>)

fun parseUnitxt(cursor: Cursor): Unitxt {
    val categoryCount = cursor.int()
    val entryCounts = cursor.intArray(categoryCount)

    val categoryEntryOffsets: List<IntArray> = entryCounts.map { entryCount ->
        cursor.intArray(entryCount)
    }

    val categories = categoryEntryOffsets.map { entryOffsets ->
        entryOffsets.map { entryOffset ->
            cursor.seekStart(entryOffset)
            cursor.stringUtf16(
                min(1024, cursor.bytesLeft),
                nullTerminated = true,
                dropRemaining = true,
            )
        }
    }

    return Unitxt(categories)
}
